File: content/snippets/cpp/interview_lcci_08_01_three_steps.cpp
#include <iostream>
using namespace std;

int waysToStep(int n) {
    const long long MOD = 1e9 + 7;
    if (n == 1) return 1;
    if (n == 2) return 2;
    if (n == 3) return 4;
    long long a = 1, b = 2, c = 4;
    for (int i = 3; i < n; i++) {
        long long d = ((a + b) % MOD + c) % MOD;
        a = b; b = c; c = d;
    }
    return (int)c;
}

int main() {
    cout << waysToStep(3) << endl; // 4
    cout << waysToStep(5) << endl; // 13
    return 0;
}

File: content/snippets/scala3/interview_lcci_08_01_three_steps.scala
object Lcci0801ThreeSteps:
    def waysToStep(n: Int): Int =
        val MOD = 1000000007L
        if (n == 1) return 1
        if (n == 2) return 2
        if (n == 3) return 4
        var a = 1L; var b = 2L; var c = 4L
        for (_ <- 3 until n) {
            val d = ((a + b) % MOD + c) % MOD
            a = b; b = c; c = d
        }
        c.toInt

    @main def main(): Unit =
        println(waysToStep(3)) // 4
        println(waysToStep(5)) // 13
